### ISSUE TITLE
Bug: [strict-boolean-expressions] Should not raise on always truthy

### ISSUE THREAD

**Issue by laverdet** [typescript-eslint/typescript-eslint#7703]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.2.2&fileType=.tsx&code=PTAEGEHsFtsg7UBLRBjSATApqArgZxQHNQBxAJwEMAHACwEUAZUAAQSwHkAzUDJcrKgAuSAG5YAUEICe1HABVZOALygA3hNCgANli5CAXKEgAjAFaChAbk2hySIrSEB%2BI-CzjyNgL6gAPuq2uvquoO6eNlr2jobG5pY%2BNhLYqNqUAqDo8PhCoKKU2rhYRopyNkg8ABT5hVgAdMFCAJSBWln4kLoNkESVAOSNfU2RoCCgAOo47lgYoEKQvEj4lCa6oABMdri6%2BMiIQrRLmZT4WACEEr5Y2qfIVTVFddFOLRptCB1d2j39z0JDPgkEjGAAMcvZhABaEyQTpYSjwSFYAAe1AE%2BEIHxByF28EguUooHwuDk5FOuUgPBBeMhuHg7lQWAx6WkkKyfBECBBdQmUywM1AMIOF10uXguGgJiw5FAqgADDZxZLpaBnM5VABGKxAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iQANOGx4ipZBWp0GzNsnxUydAEYcOSIU3oAPYtEHymqDJGWr12gC%2BIIFAA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```typescript
// Common in code using GraphQL @oneOf directive
type Type = {
  left: object;
  right?: never;
} | {
  left?: never;
  right: object;
};

declare const value: Type;
if (value.left) {
  console.log('left');
  // We need to disable 2 rules in this case!
} else if (value.right) {
  console.log('right');
}
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error",
    "@typescript-eslint/strict-boolean-expressions": "error"
  }
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

I would expect this code to only raise one error, in `no-unnecessary-condition`.

### Actual Result

Instead, `no-unnecessary-condition` raises "Unnecessary conditional, value is always truthy" and `strict-boolean-expressions` raises "Unexpected object value in conditional. The condition is always true"

### Additional Info

I believe these 2 rules are stepping on each others toes and creating redundant work. In order to suppress this warning we need to disable 2 rules instead of just 1.

`strict-boolean-expressions` should not raise issues on always truthy expressions; that is the job of `no-unnecessary-condition`. Furthermore the documentation of `no-unnecessary-condition` has this note: "`strict-boolean-expressions` - a more opinionated version of no-unnecessary-condition." I think this is misleading because it implies that `strict-boolean-expressions` is a superset, but that is not the case. Both rules are needed for completeness.

**Comment by JoshuaKGoldberg** [comment]
> ```ts
>   // We need to disable 2 rules in this case!
> } else if (value.right) {
> 

The type of `value.right` is `object`. That's truthy. So the rules are right to complain here.

This looks like an example of needing to use conditional logic such as `if`/`else` to check whether a value the type system thinks is truthy is in fact truthy. The right solution there is to fix the types (e.g. what the GraphQL tooling is auto-generating for you).
